source/volume_of_fluid/reconstruct.cc

```diff
@@ -0,0 +1,488 @@
+    const unsigned int volume_of_fluid_ind
+      = this->get_fe().component_to_system_index(volume_of_fluid_c_index, 0);
+
+    const FEVariable<dim> &volume_of_fluidN_var = field.reconstruction;
+    const unsigned int volume_of_fluidN_c_index = volume_of_fluidN_var.first_component_index;
+    const unsigned int volume_of_fluidN_blockidx = volume_of_fluidN_var.block_index;
+
+    const FEVariable<dim> &volume_of_fluidLS_var = field.level_set;
+    const unsigned int volume_of_fluidLS_c_index = volume_of_fluidLS_var.first_component_index;
+    const unsigned int n_volume_of_fluidLS_dofs = volume_of_fluidLS_var.fe->dofs_per_cell;
+    const unsigned int volume_of_fluidLS_blockidx = volume_of_fluidLS_var.block_index;
+
+    //Iterate over cells
+    for (auto cell : this->get_dof_handler().active_cell_iterators ())
+      {
+        if (!cell->is_locally_owned ())
+          continue;
+
+        double cell_volume_of_fluid;
+
+        // Obtain data for this cell and neighbors
+        cell->get_dof_indices (local_dof_indicies);
+        cell_volume_of_fluid = solution(local_dof_indicies[volume_of_fluid_ind]);
+
+        normal[0] = 0.0;
+        normal[1] = 0.0;
+        d = -1.0;
+
+        if (cell_volume_of_fluid > 1.0 - volume_fraction_threshold)
+          {
+            d = 1.0;
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = 1.0;
+          }
+        else if (cell_volume_of_fluid < volume_fraction_threshold)
+          {
+            d = -1.0;
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = 0.0;
+          }
+        else
+          {
+            initial_solution(local_dof_indicies[volume_of_fluid_ind]) = cell_volume_of_fluid;
+
+            //Identify best normal
+            // Get references to neighboring cells
+            for (unsigned int i = 0; i < 3; ++i)
+              {
+                typename DoFHandler<dim>::active_cell_iterator cen;
+                if (i == 0 || i == 2)
+                  {
+                    const unsigned int neighbor_no = (i/2);
+                    const typename DoFHandler<dim>::face_iterator face = cell->face (neighbor_no);
+                    if (face->at_boundary() && !cell->has_periodic_neighbor(neighbor_no)||
+                        face->has_children())
+                      cen = endc;
+                    else
+                      {
+                        const typename DoFHandler<dim>::cell_iterator neighbor =
+                          cell->neighbor_or_periodic_neighbor(neighbor_no);
+                        if (neighbor->level() == cell->level() &&
+                            neighbor->active())
+                          cen = neighbor;
+                        else
+                          cen = endc;
+                      }
+                  }
+                else
+                  cen = cell;
+                for (unsigned int j = 0; j < 3; ++j)
```

Comment:
Please put an empty line between the else and the for-loop otherwise this is confusing to read. Also replace 3 by an appropriate variable name. Finally, add a comment: What is `curr` below, in particular compared to `cen` above?

Reply:
Comments added to attempt clarification